File: android/app/src/main/java/com/lttprandomizer/MsuPlaylistManager.kt
package com.lttprandomizer

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.InputStream
import java.io.OutputStream
import java.time.Instant

object MsuPlaylistManager {
    private val json = Json { prettyPrint = true; ignoreUnknownKeys = true }

    fun load(inputStream: InputStream): Pair<MsuPlaylist?, String?> {
        return try {
            val text = inputStream.bufferedReader().use { it.readText() }
            val playlist = json.decodeFromString<MsuPlaylist>(text)
            if (playlist.version != 1)
                Pair(null, "Unsupported playlist version: ${playlist.version}")
            else
                Pair(playlist, null)
        } catch (ex: Exception) {
            Pair(null, "Failed to load playlist: ${ex.message}")
        }
    }

    fun save(outputStream: OutputStream, playlist: MsuPlaylist): String? {
        return try {
            val updated = playlist.copy(
                lastModified = Instant.now().toString(),
                tracks = playlist.tracks.mapValues { it.value.replace('\\', '/') }
            )
            val text = json.encodeToString(updated)
            outputStream.bufferedWriter().use { it.write(text) }
            null
        } catch (ex: Exception) {
            "Failed to save playlist: ${ex.message}"
        }
    }
}

File: windows/Services/MsuPlaylistManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using LTTPRandomizerGenerator.Models;

namespace LTTPRandomizerGenerator.Services;

public static class MsuPlaylistManager
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        PropertyNameCaseInsensitive = true
    };

    public static (MsuPlaylist? playlist, string? error) Load(string filePath)
    {
        try
        {
            string json = File.ReadAllText(filePath);
            var playlist = JsonSerializer.Deserialize<MsuPlaylist>(json, JsonOptions);
            if (playlist is null)
                return (null, "Playlist file is empty or malformed.");
            if (playlist.Version != 1)
                return (null, $"Unsupported playlist version: {playlist.Version}");
            return (playlist, null);
        }
        catch (Exception ex)
        {
            return (null, $"Failed to load playlist: {ex.Message}");
        }
    }

    public static string? Save(string filePath, MsuPlaylist playlist)
    {
        try
        {
            playlist.LastModified = DateTime.UtcNow.ToString("o");
            var normalizedTracks = playlist.Tracks
                .ToDictionary(kv => kv.Key, kv => kv.Value.Replace('\\', '/'));
            playlist.Tracks = normalizedTracks;

            string json = JsonSerializer.Serialize(playlist, JsonOptions);
            Directory.CreateDirectory(Path.GetDirectoryName(filePath)!);
            File.WriteAllText(filePath, json);
            return null;
        }
        catch (Exception ex)
        {
            return $"Failed to save playlist: {ex.Message}";
        }
    }
}
